Gems/Atom/RHI/Code/Source/RHI.Reflect/MemoryStatistics.cpp

```diff
@@ -0,0 +1,159 @@
+                poolObject.AddMember(UsedResidentInBytesAttribStr, static_cast<uint64_t>(heapMemoryUsageDevice.m_usedResidentInBytes.load()), doc.GetAllocator());
+                poolObject.AddMember(FragmentationAttribStr, heapMemoryUsageDevice.m_fragmentation, doc.GetAllocator());
+                poolObject.AddMember(UniqueAllocationsInBytesAttribStr, static_cast<uint64_t>(heapMemoryUsageDevice.m_uniqueAllocationBytes.load()), doc.GetAllocator());
+            }
+            else
+            {
+                continue;
+            }
+
+            poolObject.AddMember(BufferCountAttribStr, static_cast<uint64_t>(pool.m_buffers.size()), doc.GetAllocator());
+            poolObject.AddMember(ImageCountAttribStr,  static_cast<uint64_t>(pool.m_images.size()),  doc.GetAllocator());
+            rapidjson::Value buffersArray(rapidjson::kArrayType);
+            for (const auto& buffer : pool.m_buffers)
+            {
+                rapidjson::Value bufferObject(rapidjson::kObjectType);
+                bufferObject.AddMember(BufferNameAttribStr, rapidjson::StringRef(!buffer.m_name.IsEmpty()?buffer.m_name.GetCStr():"Unnamed Buffer"), doc.GetAllocator());
+                bufferObject.AddMember(SizeInBytesAttribStr, static_cast<uint64_t>(buffer.m_sizeInBytes), doc.GetAllocator());
+                bufferObject.AddMember(BindFlagsAttribStr, static_cast<uint32_t>(buffer.m_bindFlags), doc.GetAllocator());
+                buffersArray.PushBack(bufferObject, doc.GetAllocator());
+            }
+            poolObject.AddMember(BuffersListAttribStr, buffersArray, doc.GetAllocator());
+
+            rapidjson::Value imagesArray(rapidjson::kArrayType);
+            for (const auto& image : pool.m_images)
+            {
+                rapidjson::Value imageObject(rapidjson::kObjectType);
+                imageObject.AddMember(ImageNameAttribStr, rapidjson::StringRef(!image.m_name.IsEmpty()?image.m_name.GetCStr():"Unnamed Image"), doc.GetAllocator());
+                imageObject.AddMember(SizeInBytesAttribStr, static_cast<uint64_t>(image.m_sizeInBytes), doc.GetAllocator());
+                imageObject.AddMember(BindFlagsAttribStr, static_cast<uint32_t>(image.m_bindFlags), doc.GetAllocator());
+                imagesArray.PushBack(imageObject, doc.GetAllocator());
+            }
+            poolObject.AddMember(ImagesListAttribStr, imagesArray, doc.GetAllocator());
+            poolsArray.PushBack(poolObject, doc.GetAllocator());
+        }
+        docRoot.AddMember(PoolsAttribStr, poolsArray, doc.GetAllocator());
+        docRoot.AddMember(MemoryDataVersionMajorAttribStr, 1, doc.GetAllocator());
+        docRoot.AddMember(MemoryDataVersionMinorAttribStr, 0, doc.GetAllocator());
+        docRoot.AddMember(MemoryDataVersionRevisionAttribStr, 0, doc.GetAllocator());
+    }
+
+    void DumpPoolInfoToJson()
+    {
+            AZ::IO::Path path = AZStd::string_view(AZ::Utils::GetO3deLogsDirectory());
```

Comment:
We probably dont want to always dump on a failed allocation by default. Not sure what happens if this code runs on android/ios. Maybe hide this behind a cvar? So if a developer is trying to debug a gpu OOM crash he can enable the cvar and re-run the app.